From: coolsnowwolf <coolsnowwolf@gmail.com>
Subject: [PATCH] qca-ssdk: fix Linux 6.18 PHY driver registration

Linux 6.18 removed the single PHY driver registration helpers.
Use the array-based helpers while preserving older kernel support.

Signed-off-by: coolsnowwolf <coolsnowwolf@gmail.com>
---
--- a/src/hsl/phy/qca808x.c
+++ b/src/hsl/phy/qca808x.c
@@ -692,17 +692,25 @@
 a_int32_t qca808x_phy_driver_register(void)
 {
 	a_int32_t ret;
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+	ret = phy_drivers_register(&qca808x_phy_driver, 1, THIS_MODULE);
+#else
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,9,0))
 	ret = phy_driver_register(&qca808x_phy_driver, THIS_MODULE);
 #else
 	ret = phy_driver_register(&qca808x_phy_driver);
 #endif
+#endif
 	return ret;
 }
 
 void qca808x_phy_driver_unregister(void)
 {
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 18, 0))
+	phy_drivers_unregister(&qca808x_phy_driver, 1);
+#else
 	phy_driver_unregister(&qca808x_phy_driver);
+#endif
 }
 
 void qca808x_phydev_init(a_uint32_t dev_id, a_uint32_t port_id)
--- a/src/hsl/phy/sfp_phy.c
+++ b/src/hsl/phy/sfp_phy.c
@@ -349,7 +349,11 @@
 #if defined(IN_PHY_I2C_MODE)
 	if (hsl_port_phy_access_type_get(dev_id, port) == PHY_I2C_ACCESS) {
 		if(phydev->drv) {
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,9,0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,18,0))
+			drv = phydev->mdio.dev.driver;
+			phydrv = to_phy_driver(drv);
+			phy_drivers_unregister(phydrv, 1);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(6,9,0))
 			drv = phydev->mdio.dev.driver;
 			phydrv = to_phy_driver(drv);
 			phy_driver_unregister(phydrv);
@@ -404,7 +408,9 @@
 	int ret = 0;
 	if(sfp_phy_drv_registered == A_FALSE)
 	{
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(4,9,0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,18,0))
+		ret = phy_drivers_register(&sfp_phy_driver, 1, THIS_MODULE);
+#elif (LINUX_VERSION_CODE >= KERNEL_VERSION(4,9,0))
 		ret = phy_driver_register(&sfp_phy_driver, THIS_MODULE);
 #else
 		ret = phy_driver_register(&sfp_phy_driver);
@@ -418,7 +424,11 @@
 {
 	if (sfp_phy_drv_registered == A_TRUE)
 	{
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,18,0))
+		phy_drivers_unregister(&sfp_phy_driver, 1);
+#else
 		phy_driver_unregister(&sfp_phy_driver);
+#endif
 	}
 }
 
--- a/src/hsl/phy/hsl_phy.c
+++ b/src/hsl/phy/hsl_phy.c
@@ -1880,7 +1880,11 @@
 	rv = hsl_phy_phydev_get(dev_id, phy_addr, &phydev);
 	SW_RTN_ON_ERROR(rv);
 
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6,18,0))
+	phydev->eee_cfg.eee_enabled = (adv == 0 ? A_FALSE : A_TRUE);
+#else
 	phydev->eee_enabled = (adv == 0 ? A_FALSE : A_TRUE);
+#endif
 
 	linkmode_mod_bit(ETHTOOL_LINK_MODE_100baseT_Full_BIT,
 		phydev->advertising_eee, adv & FAL_PHY_EEE_100BASE_T);
